File: binding/include/libqcommandlineparser.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qcommandlineparser.html
pub const qcommandlineparser = struct {

    /// New constructs a new QCommandLineParser object.
    pub fn New() QtC.QCommandLineParser {
        return qtc.QCommandLineParser_new();
    }


    pub fn Tr(sourceText: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineParser_Tr(@ptrCast(sourceText));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetSingleDashWordOptionMode(self: ?*anyopaque, parsingMode: i32) void {
        qtc.QCommandLineParser_SetSingleDashWordOptionMode(@ptrCast(self), @intCast(parsingMode));
    }

    pub fn SetOptionsAfterPositionalArgumentsMode(self: ?*anyopaque, mode: i32) void {
        qtc.QCommandLineParser_SetOptionsAfterPositionalArgumentsMode(@ptrCast(self), @intCast(mode));
    }

    pub fn AddOption(self: ?*anyopaque, commandLineOption: ?*anyopaque) bool {
        return qtc.QCommandLineParser_AddOption(@ptrCast(self), @ptrCast(commandLineOption));
    }

    pub fn AddOptions(self: ?*anyopaque, options: []const u8) bool {
        return qtc.QCommandLineParser_AddOptions(@ptrCast(self), @ptrCast(options));
    }

    pub fn AddVersionOption(self: ?*anyopaque, ) QtC.QCommandLineOption {
        return qtc.QCommandLineParser_AddVersionOption(@ptrCast(self));
    }

    pub fn AddHelpOption(self: ?*anyopaque, ) QtC.QCommandLineOption {
        return qtc.QCommandLineParser_AddHelpOption(@ptrCast(self));
    }

    pub fn SetApplicationDescription(self: ?*anyopaque, description: []const u8) void {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
qtc.QCommandLineParser_SetApplicationDescription(@ptrCast(self), description_str);
    }

    pub fn ApplicationDescription(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineParser_ApplicationDescription(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.ApplicationDescription: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddPositionalArgument(self: ?*anyopaque, name: []const u8, description: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
qtc.QCommandLineParser_AddPositionalArgument(@ptrCast(self), name_str, description_str);
    }

    pub fn ClearPositionalArguments(self: ?*anyopaque, ) void {
        qtc.QCommandLineParser_ClearPositionalArguments(@ptrCast(self));
    }

    pub fn Process(self: ?*anyopaque, arguments: []const u8) void {
        qtc.QCommandLineParser_Process(@ptrCast(self), @ptrCast(arguments));
    }

    pub fn Process2(self: ?*anyopaque, app: ?*anyopaque) void {
        qtc.QCommandLineParser_Process2(@ptrCast(self), @ptrCast(app));
    }

    pub fn Parse(self: ?*anyopaque, arguments: []const u8) bool {
        return qtc.QCommandLineParser_Parse(@ptrCast(self), @ptrCast(arguments));
    }

    pub fn ErrorText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineParser_ErrorText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.ErrorText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IsSet(self: ?*anyopaque, name: []const u8) bool {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QCommandLineParser_IsSet(@ptrCast(self), name_str);
    }

    pub fn Value(self: ?*anyopaque, name: []const u8, allocator: std.mem.Allocator) []const u8 {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const _str = qtc.QCommandLineParser_Value(@ptrCast(self), name_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.Value: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Values(self: ?*anyopaque, name: []const u8) []const u8 {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QCommandLineParser_Values(@ptrCast(self), name_str);
    }

    pub fn IsSet2(self: ?*anyopaque, option: ?*anyopaque) bool {
        return qtc.QCommandLineParser_IsSet2(@ptrCast(self), @ptrCast(option));
    }

    pub fn Value2(self: ?*anyopaque, option: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineParser_Value2(@ptrCast(self), @ptrCast(option));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.Value2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Values2(self: ?*anyopaque, option: ?*anyopaque) []const u8 {
        return qtc.QCommandLineParser_Values2(@ptrCast(self), @ptrCast(option));
    }

    pub fn PositionalArguments(self: ?*anyopaque, ) []const u8 {
        return qtc.QCommandLineParser_PositionalArguments(@ptrCast(self));
    }

    pub fn OptionNames(self: ?*anyopaque, ) []const u8 {
        return qtc.QCommandLineParser_OptionNames(@ptrCast(self));
    }

    pub fn UnknownOptionNames(self: ?*anyopaque, ) []const u8 {
        return qtc.QCommandLineParser_UnknownOptionNames(@ptrCast(self));
    }

    pub fn ShowVersion(self: ?*anyopaque, ) void {
        qtc.QCommandLineParser_ShowVersion(@ptrCast(self));
    }

    pub fn ShowHelp(self: ?*anyopaque, ) void {
        qtc.QCommandLineParser_ShowHelp(@ptrCast(self));
    }

    pub fn HelpText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineParser_HelpText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.HelpText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr2(sourceText: []const u8, disambiguation: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineParser_Tr2(@ptrCast(sourceText), @ptrCast(disambiguation));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(sourceText: []const u8, disambiguation: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCommandLineParser_Tr3(@ptrCast(sourceText), @ptrCast(disambiguation), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcommandlineparser.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddPositionalArgument3(self: ?*anyopaque, name: []const u8, description: []const u8, syntax: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
const syntax_str = qtc.libqt_string{
    .len = syntax.len,
    .data = syntax.ptr,
};
qtc.QCommandLineParser_AddPositionalArgument3(@ptrCast(self), name_str, description_str, syntax_str);
    }

    pub fn ShowHelp1(self: ?*anyopaque, exitCode: i32) void {
        qtc.QCommandLineParser_ShowHelp1(@ptrCast(self), exitCode);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QCommandLineParser_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/commandlineparser.html#types
pub const enums = struct {
    pub const SingleDashWordOptionMode = enum {
        pub const ParseAsCompactedShortOptions: i32 = 0;
        pub const ParseAsLongOptions: i32 = 1;
    };

    pub const OptionsAfterPositionalArgumentsMode = enum {
        pub const ParseAsOptions: i32 = 0;
        pub const ParseAsPositionalArguments: i32 = 1;
    };

};

File: binding/src/libqcommandlineparser.cpp
#include <QCommandLineOption>
#include <QCommandLineParser>
#include <QCoreApplication>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qcommandlineparser.h>
#include "libqcommandlineparser.h"
#include "libqcommandlineparser.hxx"

QCommandLineParser* QCommandLineParser_new() {
	 return new QCommandLineParser();
}

libqt_string QCommandLineParser_Tr(const char* sourceText) {
	QString _ret = QCommandLineParser::tr(sourceText);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCommandLineParser_SetSingleDashWordOptionMode(QCommandLineParser* self, int parsingMode) {
	self->setSingleDashWordOptionMode(static_cast<QCommandLineParser::SingleDashWordOptionMode>(parsingMode));
}

void QCommandLineParser_SetOptionsAfterPositionalArgumentsMode(QCommandLineParser* self, int mode) {
	self->setOptionsAfterPositionalArgumentsMode(static_cast<QCommandLineParser::OptionsAfterPositionalArgumentsMode>(mode));
}

bool QCommandLineParser_AddOption(QCommandLineParser* self, const QCommandLineOption* commandLineOption) {
	return self->addOption(*commandLineOption);
}

bool QCommandLineParser_AddOptions(QCommandLineParser* self, const libqt_list options) {
	return self->addOptions(QList<QCommandLineOption>());
}

QCommandLineOption* QCommandLineParser_AddVersionOption(QCommandLineParser* self) {
	return new QCommandLineOption(self->addVersionOption());
}

QCommandLineOption* QCommandLineParser_AddHelpOption(QCommandLineParser* self) {
	return new QCommandLineOption(self->addHelpOption());
}

void QCommandLineParser_SetApplicationDescription(QCommandLineParser* self, const libqt_string description) {
	self->setApplicationDescription(QString::fromUtf8(description.data, description.len));
}

libqt_string QCommandLineParser_ApplicationDescription(const QCommandLineParser* self) {
	QString _ret = self->applicationDescription();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCommandLineParser_AddPositionalArgument(QCommandLineParser* self, const libqt_string name, const libqt_string description) {
	self->addPositionalArgument(QString::fromUtf8(name.data, name.len), QString::fromUtf8(description.data, description.len));
}

void QCommandLineParser_ClearPositionalArguments(QCommandLineParser* self) {
	self->clearPositionalArguments();
}

void QCommandLineParser_Process(QCommandLineParser* self, const libqt_list arguments) {
	self->process(QList<QString>());
}

void QCommandLineParser_Process2(QCommandLineParser* self, const QCoreApplication* app) {
	self->process(*app);
}

bool QCommandLineParser_Parse(QCommandLineParser* self, const libqt_list arguments) {
	return self->parse(QList<QString>());
}

libqt_string QCommandLineParser_ErrorText(const QCommandLineParser* self) {
	QString _ret = self->errorText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QCommandLineParser_IsSet(const QCommandLineParser* self, const libqt_string name) {
	return self->isSet(QString::fromUtf8(name.data, name.len));
}

libqt_string QCommandLineParser_Value(const QCommandLineParser* self, const libqt_string name) {
	QString _ret = self->value(QString::fromUtf8(name.data, name.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QCommandLineParser_Values(const QCommandLineParser* self, const libqt_string name) {
	auto _ret = self->values(QString::fromUtf8(name.data, name.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

bool QCommandLineParser_IsSet2(const QCommandLineParser* self, const QCommandLineOption* option) {
	return self->isSet(*option);
}

libqt_string QCommandLineParser_Value2(const QCommandLineParser* self, const QCommandLineOption* option) {
	QString _ret = self->value(*option);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QCommandLineParser_Values2(const QCommandLineParser* self, const QCommandLineOption* option) {
	auto _ret = self->values(*option);
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_list QCommandLineParser_PositionalArguments(const QCommandLineParser* self) {
	auto _ret = self->positionalArguments();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_list QCommandLineParser_OptionNames(const QCommandLineParser* self) {
	auto _ret = self->optionNames();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_list QCommandLineParser_UnknownOptionNames(const QCommandLineParser* self) {
	auto _ret = self->unknownOptionNames();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

void QCommandLineParser_ShowVersion(QCommandLineParser* self) {
	self->showVersion();
}

void QCommandLineParser_ShowHelp(QCommandLineParser* self) {
	self->showHelp();
}

libqt_string QCommandLineParser_HelpText(const QCommandLineParser* self) {
	QString _ret = self->helpText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QCommandLineParser_Tr2(const char* sourceText, const char* disambiguation) {
	QString _ret = QCommandLineParser::tr(sourceText, disambiguation);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QCommandLineParser_Tr3(const char* sourceText, const char* disambiguation, int n) {
	QString _ret = QCommandLineParser::tr(sourceText, disambiguation, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCommandLineParser_AddPositionalArgument3(QCommandLineParser* self, const libqt_string name, const libqt_string description, const libqt_string syntax) {
	self->addPositionalArgument(QString::fromUtf8(name.data, name.len), QString::fromUtf8(description.data, description.len), QString::fromUtf8(syntax.data, syntax.len));
}

void QCommandLineParser_ShowHelp1(QCommandLineParser* self, int exitCode) {
	self->showHelp(exitCode);
}

void QCommandLineParser_Delete(QCommandLineParser* self) {
    delete self;
}

